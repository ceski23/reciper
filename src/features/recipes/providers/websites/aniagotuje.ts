import aniagotujeIcon from 'assets/images/providers/ania_gotuje.png'
import { extractMicrodata } from 'features/recipes/providers/generic/microdata'
import type { RecipesProvider } from 'features/recipes/providers/scrapper'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode, isElementNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'

export const aniagotuje: RecipesProvider = {
	name: 'AniaGotuje.pl',
	icon: aniagotujeIcon,
	matcher: /https:\/\/aniagotuje\.pl\/.*/,
	scrape: async doc => {
		const partialRecipe = await extractMicrodata(doc)
		const color = partialRecipe.image ? await getColorFromImage(partialRecipe.image) : undefined
		const name = doc.querySelector('.article-content [itemprop="name"]')?.textContent?.trim()

		return {
			...partialRecipe,
			color,
			name,
			ingredients: parseIngredients(doc) ?? partialRecipe.ingredients,
			instructions: parseInstructions(doc) ?? partialRecipe.instructions,
			gallery: parseImages(doc),
		}
	},
}

const parseImages = (doc: Document) =>
	Array
		.from(doc.querySelectorAll('[itemprop="recipeInstructions"] img'))
		.map(elem => elem.getAttribute('data-src'))
		.filter(isDefined)

const parseIngredients = (doc: Document) => {
	const root = doc.querySelector('[itemprop="recipeInstructions"] > .ads-slot-article + div')
	const ingredientsHeaders = root?.querySelectorAll<HTMLParagraphElement>('.ing-header')

	// Single recipe
	if (ingredientsHeaders?.length === 0) {
		const recipeIngredients = Array
			.from(root?.querySelectorAll('.recipe-ing-list [itemprop="recipeIngredient"]') ?? [])
			.map(elem => elem.textContent?.trim())
			.filter(isDefined)
			.map(text => ({ text }))

		return recipeIngredients.length > 0 ? recipeIngredients : undefined
	}

	// Recipe with sub-recipes
	const recipeIngredients = Array.from(ingredientsHeaders ?? []).reduce<{
		ingredients: Recipe['ingredients']
		group?: string
	}>(({ ingredients, group }, element) => {
		const nextGroup = element.textContent?.trim() ?? group
		const nextIngredients = Array
			.from(element.nextElementSibling?.querySelectorAll('[itemprop="recipeIngredient"]') ?? [])
			.map(elem => elem.textContent?.trim())
			.filter(isDefined)
			.map(text => ({ text, group: nextGroup }))

		return {
			ingredients: ingredients.concat(nextIngredients),
			group: nextGroup,
		}
	}, {
		ingredients: [],
		group: undefined,
	}).ingredients

	return recipeIngredients.length > 0 ? recipeIngredients : undefined
}

const parseInstructions = (doc: Document) => {
	const instructions: Recipe['instructions'] = []
	let currentElement: Node | Element | null = doc.querySelector('[itemprop="recipeInstructions"] div + :is(h3, h2)')
	let header: string | undefined

	while (currentElement !== null && currentElement !== undefined) {
		if (
			isElementNode(currentElement)
			&& ['H1', 'H2', 'H3', 'H4'].includes(currentElement.tagName)
		) {
			header = currentElement.textContent?.trim()
		} else {
			const text = getTextFromNode(currentElement)

			if (text !== null && text.length > 0) {
				instructions.push({ text, group: header })
			}
		}

		currentElement = currentElement.nextSibling
	}

	return instructions.length > 0 ? instructions : undefined
}
