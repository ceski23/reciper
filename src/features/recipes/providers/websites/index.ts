export { aniagotuje } from './aniagotuje'
export { kwestiasmaku } from './kwestiasmaku'
export { wszystkiegoslodkiego } from './wszystkiegoslodkiego'
