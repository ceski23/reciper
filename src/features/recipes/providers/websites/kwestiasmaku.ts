import kwestiasmakuIcon from 'assets/images/providers/kwestia_smaku.png'
import { extractMicrodata } from 'features/recipes/providers/generic/microdata'
import type { RecipesProvider } from 'features/recipes/providers/scrapper'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'
import { parseValidNumber } from 'lib/utils/numbers'

export const kwestiasmaku: RecipesProvider = {
	name: 'Kwestia Smaku',
	icon: kwestiasmakuIcon,
	matcher: /https:\/\/www\.kwestiasmaku\.com\/.*/,
	scrape: async doc => {
		const partialRecipe = await extractMicrodata(doc)
		const color = partialRecipe?.image ? await getColorFromImage(partialRecipe.image).catch(() => undefined) : undefined
		const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim()

		const tags = Array
			.from(doc.querySelectorAll('#node-przepis-full-group-kategorie a'))
			.map(elem => elem.textContent?.toLowerCase())
			.filter(isDefined)

		const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji')
		const servings = parseValidNumber(servingsElement?.textContent)

		return {
			...partialRecipe,
			ingredients: parseIngredients(doc),
			instructions: parseInstructions(doc),
			color,
			description,
			tags,
			servings,
			gallery: parseImages(doc),
		}
	},
}

const parseIngredients = (doc: Document): Recipe['ingredients'] => {
	const ingredientsRoot = doc.querySelector('.field-name-field-skladniki')
	if (!ingredientsRoot) return []

	// Multiple ingredients groups
	if (ingredientsRoot.querySelector('.wyroznione')) {
		return Array.from(ingredientsRoot.children).reduce<{
			ingredients: Recipe['ingredients']
			group?: string
		}>(({ ingredients, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextIngredients = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				ingredients: ingredients.concat(nextIngredients),
				group: nextGroup,
			}
		}, {
			ingredients: [],
			group: undefined,
		}).ingredients
	}

	// Signle ingredients group
	const recipeIngredients = Array
		.from(ingredientsRoot.querySelectorAll('ul > li'))
		.map(elem => elem.textContent?.trim())
		.filter(isDefined)
		.map(text => ({ text }))

	return recipeIngredients
}

const parseInstructions = (doc: Document): Recipe['instructions'] => {
	const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie')
	if (!instructionsRoot) return []

	// Multiple instruction groups
	if (instructionsRoot.querySelector('.wyroznione')) {
		return Array.from(instructionsRoot.children).reduce<{
			instructions: Recipe['instructions']
			group?: string
		}>(({ instructions, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextInstructions = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				instructions: instructions.concat(nextInstructions),
				group: nextGroup,
			}
		}, {
			instructions: [],
			group: undefined,
		}).instructions
	}

	// Signle instructions group
	return Array
		.from(instructionsRoot.querySelectorAll('ul > li'))
		.map(elem => getTextFromNode(elem).trim())
		.filter(isDefined)
		.map(text => ({ text }))
}

const parseImages = (doc: Document) =>
	Array
		.from(doc.querySelectorAll('.view-zdjecia img'))
		.map(elem => elem.getAttribute('src'))
		.filter(isDefined)
