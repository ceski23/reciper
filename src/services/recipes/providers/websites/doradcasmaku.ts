import icon from 'assets/providers/doradca_smaku.png';

import { Recipe } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';

export const DoradcaSmakuProvider: Provider = (() => {
  const name = 'Doradca smaku';
  const url = 'https://www.doradcasmaku.pl/';

  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    const recipeImageElem = doc.querySelector('.recipe_item .recipe_gallery .img_wrapper img');
    const image = recipeImageElem?.getAttribute('data-src') ?? undefined;

    const partial: Partial<Recipe> = {
      image,
    };

    return Object.assign(data, partial);
  };

  return {
    name, url, scrapper, icon,
  };
})();
