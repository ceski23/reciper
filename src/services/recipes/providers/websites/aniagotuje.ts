import icon from 'assets/providers/ania_gotuje.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { colorExtractor } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { isElementNode, nonNullable } from 'utils/guards';
import { removeEmpty } from 'utils/objects';

export const AniaGotujeProvider: Provider = (() => {
  const name = 'AniaGotuje.pl';
  const url = 'https://aniagotuje.pl/';

  const parseIngredients = (doc: Document) => {
    const root = doc.querySelector('[itemprop="recipeInstructions"] > .ads-slot-article + div');
    const ingredientsHeaders = root?.querySelectorAll<HTMLParagraphElement>('.ing-header');
    const ingredients: RecipeIngredient[] = [];

    // Single recipe
    if (ingredientsHeaders?.length === 0) {
      const groupItemsElement = root?.querySelectorAll('.recipe-ing-list [itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text }));

      ingredients.push(...groupItems);

      return ingredients;
    }

    // Recipe with sub-recipes
    ingredientsHeaders?.forEach((ingHeader) => {
      const groupTitle = ingHeader.textContent?.trim();

      const groupItemsElement = ingHeader.nextElementSibling?.querySelectorAll('[itemprop="recipeIngredient"]');
      const groupItems = Array.from(groupItemsElement ?? [])
        .map((item) => item.textContent?.trim())
        .filter(nonNullable)
        .map<RecipeIngredient>((text) => ({ text, group: groupTitle }));

      ingredients.push(...groupItems);
    });

    return ingredients;
  };

  const parseInstructions = (doc: Document) => {
    const instructions: RecipeInstruction[] = [];
    let currentElement: Node | Element | null = doc.querySelector('[itemprop="recipeInstructions"] .recipe-ing-list + h2');
    let header: string | undefined;

    while (currentElement !== null && currentElement !== undefined) {
      if (
        isElementNode(currentElement)
        && ['H1', 'H2', 'H3', 'H4'].includes(currentElement.tagName)
      ) {
        header = currentElement.textContent?.trim();
      } else {
        const text = getTextFromNode(currentElement);

        if (text !== null && text.length > 0) {
          instructions.push({ text, group: header });
        }
      }

      currentElement = currentElement.nextSibling;
    }

    return instructions;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await microdataScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const recipeName = doc.querySelector('.article-content [itemprop="name"]')?.textContent?.trim();

    const partial: Partial<Recipe> = {
      color,
      name: recipeName,
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
    };

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
