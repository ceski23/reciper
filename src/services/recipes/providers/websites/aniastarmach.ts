import icon from 'assets/providers/ania_starmach.png';

import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import { colorExtractor } from 'services/recipes/providers/utils';

export const AniaStarmachProvider: Provider = (() => {
  const name = 'Ania Starmach';
  const url = 'https://aniastarmach.pl/';

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const tagsElements = doc.querySelectorAll('.tags-wrapper a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(Boolean) as string[];

    return {
      ...data,
      color,
      tags,
    };
  };

  return {
    name, url, scrapper, icon,
  };
})();
