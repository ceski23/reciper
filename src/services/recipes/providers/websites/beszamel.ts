/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/beszamel.svg';

import { Recipe, RecipeIngredient } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import jsonldScrapper from 'services/recipes/providers/jsonld';
import { colorExtractor } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import { removeEmpty } from 'utils/objects';

export const BeszamelProvider: Provider = (() => {
  const name = 'Beszamel';
  const url = 'https://beszamel.se.pl/';

  const parseIngredients = (doc: Document) => {
    const elements = doc.querySelectorAll('.ingredients__items li');
    let group: string | undefined;
    const ingredients: RecipeIngredient[] = [];

    for (const element of elements) {
      const text = getTextFromNode(element);

      if (text.endsWith(':')) {
        group = text;
      } else {
        ingredients.push(removeEmpty({ group, text }));
      }
    }

    return ingredients;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const data = await jsonldScrapper(doc);

    let color;
    if (data?.image) {
      const palette = await colorExtractor(data.image);
      color = palette.Vibrant?.hex;
    }

    const tagsElements = doc.querySelectorAll('.breadcrumb .breadcrumb__item');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase().trim())
      .filter(nonNullable);

    const partial: Partial<Recipe> = {
      color,
      tags,
      ingredients: parseIngredients(doc),
    };

    return Object.assign(data, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
