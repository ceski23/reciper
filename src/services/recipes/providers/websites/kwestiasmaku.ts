/* eslint-disable @typescript-eslint/no-loop-func */
/* eslint-disable @typescript-eslint/brace-style */
/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/kwestia_smaku.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { getColorFromImage } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import appLogger from 'utils/logger';
import { removeEmpty } from 'utils/objects';

const log = appLogger.extend('scrapper:kwestiasmaku');

export const KwestiaSmakuProvider: Provider = (() => {
  const name = 'Kwestia Smaku';
  const url = 'https://www.kwestiasmaku.com/';

  const parseIngredients = (doc: Document) => {
    const ingredientsRoot = doc.querySelector('.field-name-field-skladniki');
    if (!ingredientsRoot) return [];

    // Multiple ingredients groups
    if (ingredientsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const ingredients: RecipeIngredient[] = [];

      for (const element of ingredientsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeIngredients = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => elem.textContent?.trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          ingredients.push(...recipeIngredients);
        }
      }

      return ingredients;
    }

    // Signle ingredients group
    const ingredientsElements = ingredientsRoot.querySelectorAll('ul > li');
    const recipeIngredients = Array
      .from(ingredientsElements)
      .map((elem) => elem.textContent?.trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeIngredients;
  };

  const parseInstructions = (doc: Document) => {
    const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie');
    if (!instructionsRoot) return [];

    // Multiple instruction groups
    if (instructionsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const instructions: RecipeInstruction[] = [];

      for (const element of instructionsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeInstructions = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => getTextFromNode(elem).trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          instructions.push(...recipeInstructions);
        }
      }

      return instructions;
    }

    // Signle instructions group
    const instructionsElements = instructionsRoot.querySelectorAll('ul > li');
    const recipeInstructions = Array
      .from(instructionsElements)
      .map((elem) => getTextFromNode(elem).trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeInstructions;
  };

  const parseImages = (doc: Document) => {
    log('looking for images');

    const imageElements = doc.querySelectorAll('.view-zdjecia img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('src'))
      .filter(nonNullable);

    log('found images:', images);

    return images;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const microdata = await microdataScrapper(doc);
    const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim() || undefined;

    const tagsElements = doc.querySelectorAll('#node-przepis-full-group-kategorie a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(nonNullable);

    const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji');
    const servings = servingsElement?.textContent
      ? Number.parseInt(servingsElement.textContent, 10)
      : undefined;

    const color = microdata.image ? await getColorFromImage(microdata.image) : undefined;

    const partial: Partial<Recipe> = {
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
      color,
      description,
      tags,
      servings,
      gallery: parseImages(doc),
    };

    return Object.assign(microdata, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
