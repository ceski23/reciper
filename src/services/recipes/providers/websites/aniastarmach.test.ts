import { expect } from 'vitest';

import { lemoniada, tortUrodzinowy } from 'services/recipes/providers/fixtures/aniastarmach';
import { AniaStarmachProvider } from 'services/recipes/providers/websites/aniastarmach';

const scrapeRecipe = async (data: string) => {
  const parser = new DOMParser();
  const doc = parser.parseFromString(data, 'text/html');
  const partialRecipe = await AniaStarmachProvider.scrapper(doc);
  partialRecipe.id = '1';

  return partialRecipe;
};

describe('scrapper', () => {
  it('should scrape valid recipe', async () => {
    const recipe = await scrapeRecipe(lemoniada);
    expect(recipe).toMatchSnapshot();
  });

  it('should scrape valid recipe with ingredients groups', async () => {
    const recipe = await scrapeRecipe(tortUrodzinowy);
    expect(recipe).toMatchSnapshot();
  });
});
