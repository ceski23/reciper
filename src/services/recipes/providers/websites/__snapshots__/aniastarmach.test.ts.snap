// Vitest Snapshot v1

exports[`scrapper > should scrape valid recipe 1`] = `
{
  "color": "#7f7f7f",
  "description": "Orzeźwiająca i obłędnie cytrynowa. Lemoniada na upalne dni.",
  "id": "1",
  "image": "https://aniastarmach.pl/content/uploads/2016/05/najlepsza-lemoniada-5-1200x630.jpg",
  "ingredients": [
    {
      "text": "4 cytryny",
    },
    {
      "text": "1 pomarańcza",
    },
    {
      "text": "1 garść świeżej mięty",
    },
    {
      "text": "100 g cukru",
    },
    {
      "text": "1 l wody niegazowanej (jeśli lubisz wodę gazowaną, możesz ją wymieszać pół na pół z niegazowaną)",
    },
  ],
  "instructions": [
    {
      "text": "Z owoców wyciśnij sok, przelej przez sitko, aby pozbyć się pestek.",
    },
    {
      "text": "Miętę utrzyj z cukrem w moździerzu lub w wysokim naczyniu (wtedy użyj łyżki).",
    },
    {
      "text": "Do dzbanka przelej sok z owoców. Dodaj miętę utartą z cukrem i wodę, wymieszaj.",
    },
    {
      "text": "Lemoniada smakuje najlepiej, gdy jest schłodzona (jeśli nie masz czasu na chłodzenie, dodaj po prostu kilka kostek lodu).",
    },
  ],
  "name": "Najlepsza lemoniada",
  "prepTime": 15,
  "servings": 4,
  "tags": [
    "cytryna",
    "lemoniada",
    "mięta",
    "na słodko",
    "napój",
    "pomarańcza",
    "wegetariańskie",
  ],
}
`;

exports[`scrapper > should scrape valid recipe with ingredients groups 1`] = `
{
  "color": "#7f7f7f",
  "description": "Przepyszny tort urodzinowy, który sprawi, że ten dzień będzie naprawdę wyjątkowy :)",
  "id": "1",
  "image": "https://aniastarmach.pl/content/uploads/2016/10/tort-urodzinowy-1200x630.png",
  "ingredients": [
    {
      "group": "Biszkopt",
      "text": "6 jajek",
    },
    {
      "group": "Biszkopt",
      "text": "6 łyżek mąki ziemniaczanej",
    },
    {
      "group": "Biszkopt",
      "text": "150 g cukru",
    },
    {
      "group": "Biszkopt",
      "text": "200 g mąki pszennej",
    },
    {
      "group": "Masa",
      "text": "400 ml śmietanki kremówki 36%",
    },
    {
      "group": "Masa",
      "text": "400 g serka mascarpone",
    },
    {
      "group": "Masa",
      "text": "700 g truskawek",
    },
    {
      "group": "Masa",
      "text": "2 łyżki cukru pudru",
    },
    {
      "group": "Do nasączania biszkopta",
      "text": "0.5 szklanki wody",
    },
    {
      "group": "Do nasączania biszkopta",
      "text": "1 łyżka soku z cytryny",
    },
    {
      "group": "Do nasączania biszkopta",
      "text": "sok wyciśnięty z 2 pomarańczy",
    },
  ],
  "instructions": [
    {
      "text": "Piekarnik rozgrzej do temperatury 170 stopni.",
    },
    {
      "text": "Dno tortownicy ze zdejmowanym brzegiem o średnicy 26 cm wyłóż papierem do pieczenia. Wytnij pasek papieru i wyłóż nim brzegi tortownicy (pasek papieru powinien być wyższy od wysokości formy).",
    },
    {
      "text": "Wymieszaj mąkę pszenną z ziemniaczaną.",
    },
    {
      "text": "Oddziel białka od żółtek.",
    },
    {
      "text": "Białka ubij na sztywną pianę, pod koniec ubijania dodaj cukier, następnie pojedynczo dodawaj żółtka, wymieszaj mikserem.",
    },
    {
      "text": "Do masy powoli dosypuj mąkę, zmieszaj delikatnie za pomocą szpatułki.",
    },
    {
      "text": "Ciasto przełóż do formy.",
    },
    {
      "text": "Piecz 45 minut.",
    },
    {
      "text": "Wyciągnij blachę z piekarnika. Jak nieco ostygnie, ściągnij obręcz tortownicy. Przewróć biszkopt na drugą stronę, aby wyrównać powierzchnię (zazwyczaj po upieczeniu biszkopt ma małą górkę).",
    },
    {
      "text": "Składniki do nasączania biszkopta zagotuj, ostudź.",
    },
    {
      "text": "Śmietanę ubij z serkiem mascarpone i cukrem.",
    },
    {
      "text": "Biszkopt przetnij na dwie części, za pomocą pędzla lub łyżki nasącz oba blaty ciasta, powinny być wilgotne.",
    },
    {
      "text": "Posmaruj jeden z blatów masą śmietankową, część masy zachowaj do dekoracji. Z truskawek usuń szypułki, owoce przekrój na połówki. Na pierwszym blacie połóż połowę truskawek, przykryj drugim blatem, udekoruj śmietaną i pozostałymi owocami.",
    },
  ],
  "name": "Tort urodzinowy",
  "prepTime": 60,
  "servings": 6,
  "tags": [
    "ciasto",
    "cukier",
    "cukier puder",
    "cytryna",
    "deser",
    "dla dzieci",
    "mąka pszenna",
    "mąka ziemniaczana",
    "mascarpone",
    "na słodko",
    "owoce",
    "piknik",
    "podwieczorek",
    "przyjęcie",
    "śmietanka",
    "tort",
    "truskawki",
    "wegetariańskie",
  ],
}
`;
